### DISCUSSION TITLE
What are `in` and `out` queues and when are messages dropped?

### DISCUSSION THREAD

**Discussion by vonWojtas** [vernemq/vernemq#2045] [Q&A]
Hello again 👋 

I am still working on our whole setup and  VerneMQ is part of it.
I would like to understand it more because I am probably too dumb to understand the docs.

I would like to know what are the `in` and `out` queues I see in the exposed metrics? You can find our real life example here:
![image](https://user-images.githubusercontent.com/70715766/195561459-b7143209-3749-4711-b660-2a0667726407.png)

Also I would like to know, why does the VerneMQ start to drop messages? 
With regards to dropping, could you explain to me those settings, please?

max_inflight_messages
max_online_messages
max_offline_messages

I have set those values accordingly to the [docs](https://docs.vernemq.com/configuring-vernemq/options) and did assume that the Verne does not care about the consumer but just blasts all messages out when I set the values to **unlimitied**.

Thank you very much again in advance !

**Comment by ioolkos** [comment]
In short: message `in` is a counter that gets increased when a message enters a queue. `out` when a message leaves a queue. This gives you an aggregated view on the queue situation or even throughput. (somewhat non-intuitive, though).

Verne starts to drop messages when a queue is above its configured queue maximum (`max_online_messages` and `max_offline_messages`). An online queue only increases when you have a slow consumer. An offline queue increases because it collects messages for on offline client (that requested a persistent session before).

Let's say you have a `max_offline_messages` of 1000. This means that after any offline queue fills up with 1000 messages, Verne will drop.

`max_inflight_messages` is unrelated to queue settings. It is the maximum number of messages allowed on the wire at the outgoing end of the broker before the broker expects an acknowledgment from the consumer. It only matters for QoS > 0.

**Reply by vonWojtas** [reply]
Thank you very much for your explanation, that helps me! 

Could you tell me, why would I drop any messagen, when I have set `max_online_messages=-1` (also in conjunction with `max_inflight_messages`)
I would expect the broker to  just send out all messages to the consumer as they arrive at the broker.
My situation is that I have a single Consumer on the broker.

**Reply by ioolkos** [reply]
Your expectation seems to be correct & I cannot tell you why you see dropped messages.

**Reply by vonWojtas** [reply]
Hey 👋  
Do you have any idea what this could mean?
Is the VerneMQ overloaded or rather the client not quick enough to accept / ack messages?

**Reply by ioolkos** [reply]
@vonWojtas please understand there's too much guess work implied for me to help effectively. How do you determine you are dropping messages, what exactly are you doing etc etc.

**Reply by vonWojtas** [reply]
I understand. Thank you very much though! You helped me a lot!

**Comment by mattaylor** [comment]
To follow up on this when max_online_messages is set to `-1`, what is the expected behavior as the queue starts to fill up? Will the accumulation of queued message eventually lead to OOM errors? or will this just accumulate untill leveldb is saturated at 100%. I presume the later, in which case once leveldb is saturated will old messages then be dropped to make room for new ones.

**Comment by ioolkos** [comment]
@mattaylor `-1` will give you an unbounded queue. My guess is that you'll be seeing OOM errors first. The QoS of the published message matters too, QoS 0 messages (without message id) will be held by the queue too until consumed by the MQTT client.

LevelDB does not work as you describe in VerneMQ. That is, you cannot "saturate at 100%" and then expect LevelDB to delete old messages by itself. If at all, Verne would have to tell LevelDB what to do (and it does, in general), and after a Delete LevelDB would compact away a key/value eventually.

Since the `max_online_messages` setting serves as a buffer for online MQTT consumers, you can always observe and find a working value. The thing with `-1` is that you cannot really have unbounded queues/buffers and then still somehow restrict them.
Let me know should this create more confusion than clarity :)

**Reply by mattaylor** [reply]
I am still a little confused here. If the queue is stored in levels and leveldb is only allocated 20% of available memory, why would there by oom errors when the queue builds up indefinely

**Reply by ioolkos** [reply]
The queue state machine (`vmq_queue`) is somewhat complex, apologies for not writing up a full answer. In short: The queue is not "empty" or using no RAM just because the messages for which we need to guarantee persistence are given to the message store (LevelDB). It holds QoS 0 messages in the online queue as well as references and metadata for persisted messages as a minimum.

**Reply by mattaylor** [reply]
thanks for your help. My take away here is that having an unbounded queue is not safe, so we should set this to a high number. Is there any rule of thumb you can recommend for high hight this could be set to avoid any oom issues?

**Reply by mattaylor** [reply]
we are experiencing frequent oom related crashes from the erlang vm and there is very little in the logs or metrics to help diagnose the issue. For example on the last instance we had zero queue/dropped messages but an unbounded queue, on the previous incident we had over 500k queued messages and a small queue)
